jmxtrans-output/jmxtrans-output-kafka/src/main/java/com/googlecode/jmxtrans/model/output/kafka/DefaultResultSerializer.java

```diff
@@ -0,0 +1,113 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.googlecode.jmxtrans.model.output.kafka;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.googlecode.jmxtrans.model.Query;
+import com.googlecode.jmxtrans.model.Result;
+import com.googlecode.jmxtrans.model.Server;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static com.fasterxml.jackson.core.JsonEncoding.UTF8;
+import static com.googlecode.jmxtrans.model.naming.KeyUtils.getKeyString;
+import static com.googlecode.jmxtrans.util.NumberUtils.isNumeric;
+
+/**
+ * Original Result serializer (looks like OpenTSDB data format)
+ */
+public class DefaultResultSerializer implements ResultSerializer {
+    private static final Logger log = LoggerFactory.getLogger(DefaultResultSerializer.class);
+    private final JsonFactory jsonFactory;
+    private final ImmutableList<String> typeNames;
+    private final boolean booleanAsNumber;
+    private final String rootPrefix;
+    private final ImmutableMap<String, String> tags;
+
+    @JsonCreator
+    public DefaultResultSerializer(@JsonProperty("typeNames") List<String> typeNames,
+                                   @JsonProperty("booleanAsNumber") boolean booleanAsNumber,
+                                   @JsonProperty("rootPrefix") String rootPrefix,
+                                   @JsonProperty("tags") Map<String, String> tags) {
+        this.jsonFactory = new JsonFactory();
+        this.typeNames = ImmutableList.copyOf(typeNames);
+        this.booleanAsNumber = booleanAsNumber;
+        this.rootPrefix = rootPrefix;
+        this.tags = ImmutableMap.copyOf(tags);
+    }
+
+    @Nonnull
+    @Override
+    public Collection<String> serialize(Server server, Query query, Result result) throws IOException {
```

Comment:
Introducing a generic Json writer is a great idea! We should make it a bit more generic, by passing a json Generator as a method argument here, and letting the caller create that object. This would allow to inject a generator that is directly based on an output stream and getting rid of this intermediate String (for the generic case at least, it might not be useful for Kafka itself).

This comment might be a bit obscure, I'll try to find some time to propose actual code, which might be easier to understand. And again, if you want to have a chat on all that, just ping me.